File: converters/TET-3.0-Linux/bind/cpp/tetml.cpp
/**
 * Extract text from PDF document as XML. If an output filename is specified,
 * write the XML to the output file. Otherwise fetch the XML in memory, parse it
 * and print some information to System.out.
 * 
 * $Id: tetml.cpp,v 1.4 2008/12/23 17:31:09 rjs Exp $
 */

#include <iostream>
#include <fstream>

#include "tet.hpp"

/* Global option list. */
const string globaloptlist = "searchpath={../data "
		 "../../../resource/cmap}";

/* Document specific option list. */
const string basedocoptlist = "";

/* Page-specific option list. */
const string pageoptlist = "granularity=word";

/* Word counter for in-memory processing code. */
int word_count = 0;

/* set this to 1 to generate TETML output in memory */
int inmemory = 0;

int main(int argc, char **argv)
{
    TET tet;

    if (argc != 3)
    {
	cerr << "usage: tetml <pdffilename> <xmlfilename>" << endl;
	return(2);
    }

    try
    {
	tet.set_option(globaloptlist);
	char docoptlist[512];

	if (inmemory) {
	    sprintf(docoptlist, "tetml={} %s", basedocoptlist.c_str());
	} else {
	    sprintf(docoptlist, "tetml={filename=%s}",
			    argv[2], basedocoptlist.c_str());
	}

	if (inmemory) {
	    cout << "Processing TETML output for document \""
		    << argv[1] << "\" in memory..." << endl;
	} else {
	    cout << "Extracting TETML for document \""
		    << argv[1] << "\" to file \"" << argv[2] << "\"..." << endl;
	}
	
	int doc = tet.open_document(argv[1], docoptlist);
	if (doc == -1)
	{
	    cerr << "Error " << tet.get_errnum() << " in "
		    << tet.get_apiname() << "(): " << tet.get_errmsg() << endl;
	    return(2);
	}

	int n_pages = (int) tet.pcos_get_number(doc, "length:pages");

	/*
	 * Loop over pages in the document;
	 */
	for (int pageno = 0; pageno <= n_pages; ++pageno)
	{
	    tet.process_page(doc, pageno, pageoptlist);
	}

	/*
	 * This could be combined with the last page-related call.
	 */
	tet.process_page(doc, 0, "tetml={trailer}");

	if (inmemory)
	{
	    /* Retrieve the generated TETML data from memory. Since we have
	     * only a single call the result will contain the full TETML.
	     */
	    const char *tetml;
	    size_t len;
	    tetml = tet.get_xml_data(doc, &len, "");

	    if (tetml == NULL)
	    {
		cerr << "tetml: couldn't retrieve XML data" << endl;
		return(2);
	    }
	    ofstream ofp(argv[2]);
	    if (!ofp) {
		cerr << "tetml: couldn't open output file " << argv[2] << endl;
		return(2);
	    }
	    ofp << tetml;
	    ofp.close();
	}

	tet.close_document(doc);
    }
    catch (TET::Exception &ex) {
	cerr << "Error " << ex.get_errnum() << " in "
		<< ex.get_apiname() << "(): " << ex.get_errmsg() << endl;
    }
    return(0);
}

File: converters/TET-3.0-Linux/bind/php/tetml.php
<?php
/*
 * TET sample application for dumping PDF information in the XML language TETML
 *
 * $Id: tetml.php,v 1.5 2008/12/30 13:17:52 rp Exp $
 */

/* global option list */
$globaloptlist = "searchpath={../data ../../data ../../../resource/cmap}";

/* document-specific option list */
$basedocoptlist = "";

/* page-specific option list */
$pageoptlist = "granularity=word";

/* set this to 1 to generate TETML output in memory */
$inmemory = 0;


try {
    if ($argc !=3) {
	die("usage: tetml <pdffilename> <xmlfilename>\n");
    }

    $tet = new TET();

    $tet->set_option($globaloptlist);

    if ($inmemory) {
	$docoptlist = sprintf("tetml={} %s", $basedocoptlist);
    } else {
	$docoptlist = sprintf("tetml={filename={%s}} %s",
	    $argv[2], $basedocoptlist);
    }

    $doc = $tet->open_document($argv[1], $docoptlist);

    if ($doc == -1) {
	die(sprintf("Error %d in %s(): %s\n",
	    $tet->get_errnum(), $tet->get_apiname(), $tet->get_errmsg()));
    }

    $n_pages = $tet->pcos_get_number($doc, "length:pages");

    /* loop over pages in the document */
    for ($pageno = 1; $pageno <= $n_pages; ++$pageno) {
	$tet->process_page($doc, $pageno, $pageoptlist);
    }

    /* This could be combined with the last page-related call */
    $tet->process_page($doc, 0, "tetml={trailer}");

    if ($inmemory) {
	$fname = $argv[2];
	if (!$fp = fopen($fname, "wb")) {
	    die(sprintf("tetml: couldn't open output file '%s'\n", $fname));
	}

	/* Retrieve the generated TETML data from memory. Since we have
	 * only a single call the result will contain the full TETML.
	 */

	$tetml = $tet->get_xml_data($doc, "");
	if (!$tetml) {
	    die("tetml: couldn't retrieve XML data\n");
	}

	fwrite($fp, $tetml);
	fclose($fp);
    }

    $tet->close_document($doc);
}
catch (TETException $e) {
    die("TET exception occurred in tetml sample:\n" .
	    "[" . $e->get_errnum() . "] " . $e->get_apiname() . ": " .
	    $e->get_errmsg() . "\n");
}
catch (Exception $e) {
    die($e);
}

$tet = 0;
?>
